db-engine/src/main/java/com/splicemachine/db/impl/sql/compile/JoinNode.java

```diff
@@ -1296,19 +1304,34 @@ protected void generateCore(ActivationClassBuilder acb,
          */
         String joinResultSetString;
 
+        AccessPath ap = ((Optimizable)rightResultSet).getTrulyTheBestAccessPath();
         if (joinType==FULLOUTERJOIN) {
-            joinResultSetString=((Optimizable)rightResultSet).getTrulyTheBestAccessPath().
-                    getJoinStrategy().fullOuterJoinResultSetMethodName();
+            joinResultSetString=ap.getJoinStrategy().fullOuterJoinResultSetMethodName();
         } else if(joinType==LEFTOUTERJOIN){
-            joinResultSetString=((Optimizable)rightResultSet).getTrulyTheBestAccessPath().
-                    getJoinStrategy().halfOuterJoinResultSetMethodName();
+            joinResultSetString=ap.getJoinStrategy().halfOuterJoinResultSetMethodName();
         }else{
-            joinResultSetString=((Optimizable)rightResultSet).getTrulyTheBestAccessPath().
-                    getJoinStrategy().joinResultSetMethodName();
+            joinResultSetString=ap.getJoinStrategy().joinResultSetMethodName();
         }
 
         acb.pushGetResultSetFactoryExpression(mb);
         int nargs=getJoinArguments(acb,mb,joinClause);
+        if (RSUtils.isMJ(ap)) {
+            nargs++;
+
+            CompilerContext.NewMergeJoinExecutionType
+                newMergeJoin = getCompilerContext().getNewMergeJoin();
+
+            // Favor Old Merge Join if we estimate probing won't reduce the scanned row count much.
+            boolean chooseOldMergeJoin = favorOldMergeJoin(ap.getCostEstimate());
+            if (chooseOldMergeJoin) {
+                if (newMergeJoin == SYSTEM)
+                    newMergeJoin = SYSTEM_OFF;
+                else if (newMergeJoin == ON)
+                    newMergeJoin = OFF;
+            }
```

Comment:
else throw?

Reply:
The values OFF and FORCED don't need to be checked because they can't be overridden.  I added a check for SYSTEM_OFF, because that setting is not expected:
`                else if (newMergeJoin == SYSTEM_OFF) {
                    // SYSTEM_OFF should not be allowed as a setting of the property.
                    // It is only used for communication from the parser to the
                    // execution engine.
                    if(SanityManager.DEBUG)
                        SanityManager.THROWASSERT(
                                "Illegal setting of newMergeJoin in "+this.getClass().getName()+this);
                }`

I chose not to do a plain throw that would be hit on a cluster because this isn't a critical condition, and it would be good if we still allowed the query to run.  We don't have any locations where the property is allowed to be set to SYSTEM_OFF, so I don't anticipate any issues.